src/java.base/share/native/libjli/java.c

```diff
@@ -387,83 +387,97 @@ JLI_Launch(int argc, char ** argv,              /* main argc, argv */
-    CHECK_EXCEPTION_FAIL();
-    jobject mainObject = (*env)->NewObject(env, mainClass, constructor);
-    CHECK_EXCEPTION_NULL_FAIL(mainObject);
     jmethodID mainID = (*env)->GetMethodID(env, mainClass, "main",
                                  "([Ljava/lang/String;)V");
-    CHECK_EXCEPTION_FAIL();
+    CHECK_EXCEPTION_CONTINUE_OR_FAIL();
+    jmethodID constructor = (*env)->GetMethodID(env, mainClass, "<init>", "()V");
+    CHECK_EXCEPTION_CONTINUE_OR_FAIL();
+    jobject mainObject = (*env)->NewObject(env, mainClass, constructor);
+    if (mainObject == NULL) {
+        // new instance construction failed, don't call the main method,
+        // and don't continue with the next variant;
+        // leave any exception pending, so that it is visible to the caller:
+        return 0;
+    }
     (*env)->CallVoidMethod(env, mainObject, mainID, mainArgs);
-    return 1;
+    return 0;
  }
 
 /*
- * Invoke a static main without arguments. Returns 1 (true) if successful otherwise
- * processes the pending exception from GetStaticMethodID and returns 0 (false).
+ * Invoke a static main without arguments.  Returns 1 (true) if processing should
+ * continue with another main variant, returns 0 (false) otherwise. An exception
+ * may be pending in either case.
  */
 int
 invokeStaticMainWithoutArgs(JNIEnv *env, jclass mainClass) {
     jmethodID mainID = (*env)->GetStaticMethodID(env, mainClass, "main",
                                        "()V");
-    CHECK_EXCEPTION_FAIL();
+    CHECK_EXCEPTION_CONTINUE_OR_FAIL();
     (*env)->CallStaticVoidMethod(env, mainClass, mainID);
-    return 1;
+    return 0;
 }
 
 /*
- * Invoke an instance main without arguments. Returns 1 (true) if successful otherwise
- * processes the pending exception from GetMethodID and returns 0 (false).
+ * Invoke an instance main without arguments.  Returns 1 (true) if processing should
+ * continue with another main variant, returns 0 (false) otherwise. An exception
+ * may be pending in either case.
  */
 int
 invokeInstanceMainWithoutArgs(JNIEnv *env, jclass mainClass) {
-    jmethodID constructor = (*env)->GetMethodID(env, mainClass, "<init>", "()V");
-    CHECK_EXCEPTION_FAIL();
-    jobject mainObject = (*env)->NewObject(env, mainClass, constructor);
-    CHECK_EXCEPTION_NULL_FAIL(mainObject);
     jmethodID mainID = (*env)->GetMethodID(env, mainClass, "main",
                                  "()V");
-    CHECK_EXCEPTION_FAIL();
+    CHECK_EXCEPTION_CONTINUE_OR_FAIL();
+    jmethodID constructor = (*env)->GetMethodID(env, mainClass, "<init>", "()V");
+    CHECK_EXCEPTION_CONTINUE_OR_FAIL();
+    jobject mainObject = (*env)->NewObject(env, mainClass, constructor);
+    if (mainObject == NULL) {
+        // new instance construction failed, don't call the main method,
+        // and don't continue with the next variant;
+        // leave any exception pending, so that it is visible to the caller:
+        return 1;
```

Comment:
Hello Jan, the comment says "don't continue", yet this returns `1` which represents "continue". Should this return `0` instead?

Reply:
Uh, you are right of course. Thanks for noticing! Fixed. (I believe the difference here is not currently observable, as this is the last step in the cascade, but may become observable if the order changes.)